### PR TITLE
Add support for whitelisting IP ranges

### PR DESCRIPTION
**Pull request by mkjpryor-stfc**
This PR adds support for whitelisting IP ranges. This is not useful for the main `nip.io` site, but may be useful for people who want to use this backend to run a custom wildcard DNS service. As such, I hope you will consider it for inclusion.

My use case is that I need to run a custom wildcard DNS service for a domain that I own in order to provide DNS entries for a service that I run (mainly to avoid being hobbled by Let's Encrypt rate limiting). However it is not acceptable to allow any IP address in the world to have a DNS entry under this domain for reputational reasons, so I need to be able to restrict the IPs that produce a successful response to those in the IP range for my service.

This PR allows me to do this.

### CODE REVIEW

**Comment by mkjpryor-stfc** [comment]
@rses @vanillajonathan 

Is this PR something that you might consider?

I also refactored the IP address validation to use `ipaddress` from the Python standard library.

File: README.md [hunk 1]
```diff
@@ -30,8 +30,11 @@ environment variables override those:
 
 `NIPIO_SOA_NS`: SOA name server.
 
-`NIPIO_NAMESERVERS`: A space seperated list of domain=ip nameserver pairs. Example: `ns1.nip.io=127.0.0.1 ns2.nip.io=127.0.0.1`.
+`NIPIO_NAMESERVERS`: A space separated list of domain=ip nameserver pairs. Example: `ns1.nip.io=127.0.0.1 ns2.nip.io=127.0.0.1`.
 
-`NIPIO_BLACKLIST`: A space seperated list of description=ip blacklisted pairs. Example: `some_description=10.0.0.1 other_description=10.0.0.2`.
+`NIPIO_WHITELIST`: A space separated list of description=range pairs for whitelisted ranges in CIDR format.
```
Comment (vanillajonathan) [inline README.md:35]: "A space separated list" -> "A space-separated list" (with hyphen)

**Comment by mkjpryor-stfc** [comment]
@vanillajonathan Do you have any more comments to add to your review or are you otherwise happy?

File: README.md [hunk 2]
```diff
@@ -30,8 +30,11 @@ environment variables override those:
...
+An IP must be in one of the whitelisted ranges for a response to be returned. Example: `whitelist1=192.168.0.0/16 whitelist2=127.0.0.0/8`.
```
Comment (vanillajonathan) [inline README.md:36]: "An IP must be" -> "An IP address must be"

File: nipio/backend.py [hunk 3]
```diff
@@ -239,8 +248,17 @@ def handle_unknown(self, qtype: str, qname: str) -> None:
         _write('LOG', f'Unknown type: {qtype}, domain: {qname}')
         _write('END')
 
-    def handle_blacklisted(self, ip_address: str) -> None:
-        _write('LOG', f'Blacklisted: {ip_address}')
+    def is_permitted(self, ip_address: IPv4Address) -> bool:
```
Comment (vanillajonathan) [inline nipio/backend.py:251]: Perhaps ought to be named _is_permitted, prefixed with an underscore?

Comment (mkjpryor-stfc) [reply nipio/backend.py:251]: I wasn’t sure about this as none of the `handle_*` methods have a leading `_`. 

Personally I think it should be treated as part of the public API. If I was someone who was extending this class, it is a possible extension point that I might want to override.

**Comment by vanillajonathan** [comment]
Just by taking a quick glance at it, it looks good, but what do I know.

The PR also does more than one thing, not only does it add a whitelist feature, it also changes the path from `/usr/local/bin` to `/opt/nip.io`. Don't know how @rses feel about that. Usually its best that one PR does one thing.

I am not a maintainer so I can't merge. Last I heard from rses, he was a bit busy.

File: nipio/backend.py (hunk 3 above)
Comment (mkjpryor-stfc) [reply nipio/backend.py:251]: That is not a strong opinion though. Happy to change it if you prefer.

**Comment by mkjpryor-stfc** [comment]
@vanillajonathan 

I’m not that bothered about the change to `/opt/nip.io‘ - just makes more sense to me. I would happily revert the changes to the Dockerfile if it was a blocker.

**Comment by rses** [comment]
Sorry for the late reply. I'll look at the outstanding PRs this weekend (promise!). I had a few yaks to shave the past 2-3 weeks.

File: nipio/backend.py (hunk 3 above)
Comment (rses) [reply nipio/backend.py:251]: This should be `_is_permitted` - it is an implementation detail that is closed to this class.

File: nipio_tests/backend_test.py [hunk 4]
```diff
@@ -60,7 +61,87 @@ def test_backend_ends_response_to_A_request_if_ip_is_blacklisted(self):
...
+        )
+
+    def test_backend_with_empty_whitelist_responds_to_A_request_for_valid_ip(self):
+        self._send_commands(
+            ["Q", "subdomain.10.0.10.1.lcl.io", "IN", "A", "1", "127.0.0.1"]
+        )
+
+        backend = self._create_backend()
+        backend.whitelisted_ranges = []
+        backend.run()
```
Comment (rses) [inline nipio_tests/backend_test.py:122]: Can we extract this out to a method called `_run_backend_without_whitelist` and above as well?

**Review by rses — CHANGES_REQUESTED** [review]

**Comment by rses** [comment]
The change to `/opt/nip.io` is fine - as long as it works, I don't mind where it lives :rofl:  `/opt/nip.io` seems more sensible

**Comment by vanillajonathan** [comment]
@mkjpryor-stfc There were some changes requested by @rses, can you do the changes so he can merge?

**Comment by vanillajonathan** [comment]
@mkjpryor-stfc ping!

**Comment by vanillajonathan** [comment]
@mkjpryor-stfc ping!

**Comment by vanillajonathan** [comment]
@mkjpryor-stfc Can you perform the requested changes by @rses?

**Comment by rses** [comment]
I've merged this feature and made some cleanup:

https://github.com/exentriquesolutions/nip.io/commit/c7836839af4bee4672f40b856bea1ec43e0c6560

Thanks for adding this feature in! It's pretty useful. :smile:

**Pull request closed without merging**